### PR TITLE
Introduction to Prisma with Docker

### LINKED ISSUE #4322: What is Prisma - How  Prisma  Works - Create a Prisma server - Run a Prisma server with Docker
#### Proposed title of article

What is Prisma? How to run a Prisma server with Docker containers

#### Proposed article introduction

At some point, you will need a database for your back-end application, be it with python, java, Node.js, etc. This is where a connection to the database is needed. For the backed to communicate with the database of your choice, you need a way to connect your server to that database. So when someone sends a request/mutation to the server, the server modifies that data and writes it to the database. When someone writes a query requesting data, the server reads that query and sends it to the database. This is opposed to just reading from one of our static arrays.

Now to connect to that database you can use native divers. Native divers are mainly libraries that allow you to connect to a database. For example, when working with Node.js python and a MySQL database you can choose to use a MySQL native drive provided by the Node.js NPM. However, they do not provide a way to modify or validate the data structures or provide a way to model the relationships between the databases. This means you, can't model your data relations, seed, or migrate it. With native drivers, the database connection is natively implemented using the barebones of the programming language being used.

Other options would be to use ORM (object-relational mapping). For example, when working with Node.js Mangodb provide a mongoose (an npm library) database, driver. this makes it easy for a developer to connect a node.js backend to a MongoDB database. ORM will introduce you to some new features. This includes data modeling, data validation, migrating data, describe the relationships between different data fields, etc. For example, if you are using an SQL database you can model your schema and the fields you want to use. you can validate your data model, for example by saying a password field needs to have at least six characters, a certain field need to be integers, etc. manipulating this data model becomes easier as you can be adding and removing fields as the data structure changes over time.

Prisma falls under the category of an ORM and it has all features of a typical ORM such as mongoose and sequelize for Node.js. Unlike the other ORM, Prisma comes with a tone of other added features. For example, Prisma is database agnostic. Unlike the other sequelize, mongoose will only work on a MongoDB database and sequelize will only work on SQL-based databases. If you decide you want to switch from MongoDB to PostgreSQL you have to rewrite most of your applications since the two libraries are so different. With Prisma, we don't have to do that Prisma [supports major database](https://www.prisma.io/docs/reference/database-reference/supported-databases) such as MySQL, PostgreSQL, MongoDB, MariaDB, Azure SQL, Microsoft SQL Server AWS Aurora, and AWS Aurora Serverless. This means that you can easily pick the database that fits the code structure of your application and even switch between databases without the need to change any code structure of your application.

In this tutorial, we will learn more about Prisma and build a simple server that accesses Prisma to read its database schema, then run it on a Docker container.

#### Key takeaways 

What is Prisma
How Prisma works
The concept of Prisma 
Build a server that uses Prisma
Run the Prisma server with Docker

#### Article quality
This will be an absolute beginner guide that will explain the concept of Prisma, how it works and how r run a Prisma server.

Most tutorials just explain the implementation of Prisma to other languages but never provide a detailed guide of what Prisma is. Also, with the advanced trend of Docker, it would be fine to run Prisma with Docker. For example, when running Node.js with Docker, a developer can additionally choose to use Docker. Subsequently, the developer needs to know to run Prisma with Docker so that Node.js can access it.

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Faithdroid**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #4322

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by srishilesh** [comment]
I'll complete my preliminary review by tomorrow.

**Comment by WanjaMIKE** [comment]
There appears to be several sentences taken from other sources. Please correct them. Thanks
[New Scan 5_19 PM Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/8126522/New.Scan.5_19.PM.Copyleaks.Report.pdf)

File: content/articles/dockerized-prisma-postgres-api/index.md [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+
+Prisma falls under the category of an ORM, and it has all features of a typical ORM, such as mongoose and sequelize. Unlike the other ORM, Prisma comes with a tone of other added features. Prisma is database agnostic. Mongoose only works on a MongoDB database, and sequelize only works on SQL-based databases. If you decide to switch from MongoDB to PostgreSQL, you have to rewrite most of your applications since the two libraries are so different. You don't have to do that with Prisma. Prisma [supports major database](https://www.prisma.io/docs/reference/database-reference/supported-databases) such as MySQL, PostgreSQL, MongoDB, MariaDB, Azure SQL, Microsoft SQL Server AWS Aurora, and AWS Aurora Serverless. This means that you can easily pick the database that fits the code structure of your application and even switch between databases without the need to change any code structure of your application.
```
Comment (srishilesh) [inline content/articles/dockerized-prisma-postgres-api/index.md:7]: Can you break down these paragraphs into smaller ones? One paragraph must talk about one idea.

Also, add necessary reference links. For example, for ORM, MongoDB, AWS Aurora and so on.

File: content/articles/dockerized-prisma-postgres-api/index.md [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+
+This tutorial will learn more about the Prisma ecosystem and build a simple server that accesses Prisma to read its database schema and then run it on a Docker container.
+
```
Comment (srishilesh) [inline content/articles/dockerized-prisma-postgres-api/index.md:10]: Can you add a `Table of contents` and `Prerequisities` sections

File: content/articles/dockerized-prisma-postgres-api/index.md [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+### Getting started
+
+Prisma allows you to write type-safe database schemas. When setting a database, everything needs to be well set to define what that database will look like. In this case, Prisma abstracts you from writing SQL queries. Thus you have to make sure you write safe database access schemas.
+
+To have these safe schemas:
+
+1. Ensure you have Prisma installed on your computer.
```
Comment (srishilesh) [inline content/articles/dockerized-prisma-postgres-api/index.md:17]: Add steps to install them. Also, add reference link

File: content/articles/dockerized-prisma-postgres-api/index.md [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+2. Use a text editor to write this schema.
+3. Ensure you have Prisma installed in your text editor to help you write type-safe Prisma codes.
+
+This tutorial will use [Visual Studio Code](https://code.visualstudio.com/). Once you have it installed, ensure you have the [Prisma extention](https://marketplace.visualstudio.com/items?itemName=Prisma.prisma) installed. Prisma extension adds syntax highlighting, formatting, auto-completion, jump-to-definition, and linting for `.prisma files`. This helps you avoid errors by providing a reliable and utterly type-safe API built exclusively for your Prisma code.
+
+Prisma provides [Prisma studio](https://www.prisma.io/studio), allowing you to interact with `.prisma files`. Ensure you have it installed on your computer. Also, ensure Node.js is (at least 64 bits) installed. This guide will use Node.js JavaScript runtime to set up the Prisma schemas.
+
+This guide also runs Prisma using Docker. Ensure you have [Docker](https://www.docker.com/products/docker-desktop) installed on your computer and have some basic knowledge of running and using Docker. This [guide](/engineering-education/docker-concepts/) will help you get started with Docker.
```
Comment (srishilesh) [inline content/articles/dockerized-prisma-postgres-api/index.md:25]: You can move these to the pre-requisities section

File: content/articles/dockerized-prisma-postgres-api/index.md [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+
+### The main components of Prisma
+
+The Prisma ecosystem is made up of three major components. These are;
+
+- Prisma migrate
```
Comment (srishilesh) [inline content/articles/dockerized-prisma-postgres-api/index.md:31]:
```suggestion
#### Prisma migrate
```

Comment (srishilesh) [reply content/articles/dockerized-prisma-postgres-api/index.md:31]: Do the same for the remaining 2 sub-topics

File: content/articles/dockerized-prisma-postgres-api/index.md [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+
+Your basic server is ready. Let's now wrap the whole API on Docker.
+
+### Run the Prisma server with Docker and Docker-compose
+
+First, create Docker file environments. This allows you to set the command necessary to run your code on a Docker container. Go ahead and create a `Dockerfile` at your project's root. This file will have the following commands.
+
+- Add the Node.js Docker image
+
+```bash
```
Comment (srishilesh) [inline content/articles/dockerized-prisma-postgres-api/index.md:341]: Isn't it `dockerfile`?

File: content/articles/dockerized-prisma-postgres-api/index.md [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+
+- Create a directory that runs the app on Docker
+
+```bash
+WORKDIR /app
+```
+
+- Add a COPY command to copy the project files to the Docker `/app` directory
+
+```bash
```
Comment (srishilesh) [inline content/articles/dockerized-prisma-postgres-api/index.md:353]: Isn't it `dockerfile`?

File: content/articles/dockerized-prisma-postgres-api/index.md [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+![post-response](/engineering-education/dockerized-prisma-postgres-api/post-response.png)
+
+Finally, you can check this new todo using the GET `http://localhost:3000/api/todos`, as shown below:
+
+![post-api](/engineering-education/dockerized-prisma-postgres-api/post-api.png)
+
+### Conclusion
+
+This guide helped you run more about Prisma. In the end, you created a basic API using the PostgreSQL database. Then run it via Docker containers. Docker has enabled us to create the whole API without configuring PostgreSQL and Node.js on the local computer.
+
```
Comment (srishilesh) [inline content/articles/dockerized-prisma-postgres-api/index.md:515]: Can you add a GitHub URL containing your source code?

Also, add a `Further reading` section that contains reference links

**Review by srishilesh — CHANGES_REQUESTED** [review]
Hey @Faithdroid ,

Thanks for submitting this PR.

Please carefully go through the minor suggestions that I've made. Also, make sure to recheck the plagiarism report.

Ping me when you're done with the changes.

Thanks.

**Comment by Faithdroid** [comment]
Hello, check this @srishilesh

**Comment by srishilesh** [comment]
Hey @Faithdroid ,

Thanks for making the changes.

Let me start my tech review. Will keep you posted about it.

Thanks.

**Comment by srishilesh** [comment]
I'll complete my final review by tomorrow.

**Comment by srishilesh** [comment]
Starting my final review

**Comment by srishilesh** [comment]
I've reviewed almost 40% of the article. It is longer than usual, so it's taking time.

I'll make sure to complete it by tomorrow.

**Review by srishilesh — APPROVED** [review]
Hey @Faithdroid ,

Great article.

Make sure to go through the changes I've made.

Moving for the final review.

Thanks.

**Comment by hectorkambow** [comment]
jumping on this one

**Pull request merged**